File: flysweeper/minesweeper.py
"""Plain Minesweeper. Nothing here knows about flies."""
from __future__ import annotations

from collections import deque

import numpy as np

HIDDEN = -1
FLAG = 9
MINE = 10  # shown only after the game is lost

NEIGHBORS = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]


class Minesweeper:
    def __init__(self, rows: int = 9, cols: int = 9, mines: int = 10, seed: int = 0):
        if mines >= rows * cols - 9:
            raise ValueError("too many mines for a safe first click")
        self.rows, self.cols, self.n_mines = rows, cols, mines
        self.rng = np.random.default_rng(seed)
        self.seed = seed
        self.mines = np.zeros((rows, cols), dtype=bool)
        self.numbers = np.zeros((rows, cols), dtype=np.int8)
        self.revealed = np.zeros((rows, cols), dtype=bool)
        self.flagged = np.zeros((rows, cols), dtype=bool)
        self.placed = False
        self.lost = False
        self.moves = 0
        self.mine_hit: tuple[int, int] | None = None

    # ---- state -------------------------------------------------------------------------
    @property
    def total_safe(self) -> int:
        return self.rows * self.cols - self.n_mines

    @property
    def safe_revealed(self) -> int:
        return int(self.revealed.sum())

    @property
    def won(self) -> bool:
        return not self.lost and self.placed and self.safe_revealed == self.total_safe

    @property
    def over(self) -> bool:
        return self.lost or self.won

    @property
    def cleared_fraction(self) -> float:
        return self.safe_revealed / self.total_safe

    def in_bounds(self, r: int, c: int) -> bool:
        return 0 <= r < self.rows and 0 <= c < self.cols

    def neighbors(self, r: int, c: int):
        for dr, dc in NEIGHBORS:
            rr, cc = r + dr, c + dc
            if self.in_bounds(rr, cc):
                yield rr, cc

    def visible(self) -> np.ndarray:
        """int8 grid: -1 hidden, 0-8 revealed number, 9 flag, 10 exploded/revealed mine."""
        v = np.full((self.rows, self.cols), HIDDEN, dtype=np.int8)
        v[self.revealed] = self.numbers[self.revealed]
        v[self.flagged & ~self.revealed] = FLAG
        if self.lost:
            v[self.mines] = MINE
        return v

    # ---- moves -------------------------------------------------------------------------
    def _place_mines(self, r0: int, c0: int) -> None:
        forbidden = {(r0, c0), *self.neighbors(r0, c0)}
        candidates = [(r, c) for r in range(self.rows) for c in range(self.cols) if (r, c) not in forbidden]
        pick = self.rng.choice(len(candidates), size=self.n_mines, replace=False)
        for k in pick:
            self.mines[candidates[k]] = True
        for r in range(self.rows):
            for c in range(self.cols):
                self.numbers[r, c] = sum(self.mines[rr, cc] for rr, cc in self.neighbors(r, c))
        self.placed = True

    def reveal(self, r: int, c: int) -> tuple[str, int]:
        """Returns (outcome, newly_revealed). outcome in {'mine', 'safe', 'noop'}."""
        if self.over or not self.in_bounds(r, c) or self.revealed[r, c] or self.flagged[r, c]:
            return "noop", 0
        self.moves += 1
        if not self.placed:
            self._place_mines(r, c)
        if self.mines[r, c]:
            self.lost = True
            self.mine_hit = (r, c)
            return "mine", 0
        before = self.safe_revealed
        q = deque([(r, c)])
        self.revealed[r, c] = True
        while q:
            rr, cc = q.popleft()
            if self.numbers[rr, cc] == 0:
                for nr, nc in self.neighbors(rr, cc):
                    if not self.revealed[nr, nc] and not self.mines[nr, nc]:
                        self.revealed[nr, nc] = True
                        self.flagged[nr, nc] = False
                        q.append((nr, nc))
        return "safe", self.safe_revealed - before

    def toggle_flag(self, r: int, c: int) -> bool:
        if self.over or not self.in_bounds(r, c) or self.revealed[r, c]:
            return False
        self.moves += 1
        self.flagged[r, c] = not self.flagged[r, c]
        return True

    def hidden_cells(self) -> list[tuple[int, int]]:
        return [(r, c) for r in range(self.rows) for c in range(self.cols) if not self.revealed[r, c] and not self.flagged[r, c]]


def solver_move(game: Minesweeper, rng: np.random.Generator) -> tuple[str, int, int]:
    """Single-point Minesweeper logic; guesses when stuck. Returns (kind, r, c)."""
    if not game.placed:
        return "reveal", game.rows // 2, game.cols // 2
    vis = game.visible()
    safe, mines = set(), set()
    for r in range(game.rows):
        for c in range(game.cols):
            n = vis[r, c]
            if n < 0 or n > 8:
                continue
            hidden = [(rr, cc) for rr, cc in game.neighbors(r, c) if not game.revealed[rr, cc]]
            flags = [p for p in hidden if game.flagged[p]]
            unknown = [p for p in hidden if not game.flagged[p]]
            if not unknown:
                continue
            if n - len(flags) == len(unknown):
                mines.update(unknown)
            elif n == len(flags):
                safe.update(unknown)
    if safe:
        r, c = sorted(safe)[rng.integers(len(safe))]
        return "reveal", r, c
    if mines:
        r, c = sorted(mines)[rng.integers(len(mines))]
        return "flag", r, c
    hidden = game.hidden_cells()
    if not hidden:
        return "noop", 0, 0
    # Prefer cells not adjacent to any number (no constraint information).
    frontier = {p for r in range(game.rows) for c in range(game.cols) if 0 <= vis[r, c] <= 8 for p in game.neighbors(r, c)}
    unconstrained = [p for p in hidden if p not in frontier]
    pool = unconstrained or hidden
    r, c = pool[rng.integers(len(pool))]
    return "reveal", r, c

File: web/js/minesweeper.js
// Plain Minesweeper, a port of flysweeper/minesweeper.py. Nothing here knows about flies.
//
// Visible codes: -1 hidden, 0-8 revealed number, 9 flag, 10 mine (shown after losing).
// Mines are either placed on the first reveal (classic safe first click), or supplied as a
// pre-generated layout so two boards can share the exact same mines.

import { RNG } from './rng.js';

export const HIDDEN = -1;
export const FLAG = 9;
export const MINE = 10;

export const NEIGHBORS = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]];

export class Minesweeper {
  /**
   * @param {number} rows
   * @param {number} cols
   * @param {number} nMines
   * @param {object} [opts]
   * @param {number} [opts.seed]          seed for mine placement (classic mode)
   * @param {Uint8Array} [opts.layout]    pre-generated mines (rows*cols, 1 = mine); board starts "placed"
   */
  constructor(rows = 9, cols = 9, nMines = 10, opts = {}) {
    if (nMines >= rows * cols - 9) throw new Error('too many mines for a safe first click');
    this.rows = rows; this.cols = cols; this.nMines = nMines;
    this.seed = opts.seed ?? 0;
    this.rng = new RNG(this.seed);
    this.mines = new Uint8Array(rows * cols);
    this.numbers = new Int8Array(rows * cols);
    this.revealed = new Uint8Array(rows * cols);
    this.flagged = new Uint8Array(rows * cols);
    this.placed = false;
    this.lost = false;
    this.moves = 0;
    this.mineHit = null;
    if (opts.layout) {
      this.mines.set(opts.layout);
      this._computeNumbers();
      this.placed = true;
    }
  }

  /** A mine layout with a guaranteed mine-free 3x3 zone around (safeR, safeC), like _place_mines. */
  static makeLayout(rows, cols, nMines, safeR, safeC, rng) {
    const forbidden = new Set([safeR * cols + safeC]);
    for (const [dr, dc] of NEIGHBORS) {
      const r = safeR + dr, c = safeC + dc;
      if (r >= 0 && r < rows && c >= 0 && c < cols) forbidden.add(r * cols + c);
    }
    const candidates = [];
    for (let i = 0; i < rows * cols; i++) if (!forbidden.has(i)) candidates.push(i);
    const layout = new Uint8Array(rows * cols);
    for (const k of rng.sample(candidates.length, nMines)) layout[candidates[k]] = 1;
    return layout;
  }

  get totalSafe() { return this.rows * this.cols - this.nMines; }
  get safeRevealed() { let s = 0; for (let i = 0; i < this.revealed.length; i++) s += this.revealed[i]; return s; }
  get won() { return !this.lost && this.placed && this.safeRevealed === this.totalSafe; }
  get over() { return this.lost || this.won; }
  get clearedFraction() { return this.safeRevealed / this.totalSafe; }
  get flagsPlaced() { let s = 0; for (let i = 0; i < this.flagged.length; i++) s += this.flagged[i] && !this.revealed[i] ? 1 : 0; return s; }

  inBounds(r, c) { return r >= 0 && r < this.rows && c >= 0 && c < this.cols; }

  *neighbors(r, c) {
    for (const [dr, dc] of NEIGHBORS) {
      const rr = r + dr, cc = c + dc;
      if (this.inBounds(rr, cc)) yield [rr, cc];
    }
  }

  /** Int8Array(rows*cols): -1 hidden, 0-8 number, 9 flag, 10 mine (after losing). */
  visible() {
    const v = new Int8Array(this.rows * this.cols).fill(HIDDEN);
    for (let i = 0; i < v.length; i++) {
      if (this.revealed[i]) v[i] = this.numbers[i];
      else if (this.flagged[i]) v[i] = FLAG;
    }
    if (this.lost) for (let i = 0; i < v.length; i++) if (this.mines[i]) v[i] = MINE;
    return v;
  }

  _computeNumbers() {
    for (let r = 0; r < this.rows; r++) for (let c = 0; c < this.cols; c++) {
      let n = 0;
      for (const [rr, cc] of this.neighbors(r, c)) n += this.mines[rr * this.cols + cc];
      this.numbers[r * this.cols + c] = n;
    }
  }

  _placeMines(r0, c0) {
    this.mines.set(Minesweeper.makeLayout(this.rows, this.cols, this.nMines, r0, c0, this.rng));
    this._computeNumbers();
    this.placed = true;
  }

  /** Returns {result: 'mine'|'safe'|'noop', newly: number}. */
  reveal(r, c) {
    const i = r * this.cols + c;
    if (this.over || !this.inBounds(r, c) || this.revealed[i] || this.flagged[i]) return { result: 'noop', newly: 0 };
    this.moves += 1;
    if (!this.placed) this._placeMines(r, c);
    if (this.mines[i]) {
      this.lost = true;
      this.mineHit = [r, c];
      return { result: 'mine', newly: 0 };
    }
    const before = this.safeRevealed;
    const q = [[r, c]];
    this.revealed[i] = 1;
    while (q.length) {
      const [rr, cc] = q.shift();
      if (this.numbers[rr * this.cols + cc] === 0) {
        for (const [nr, nc] of this.neighbors(rr, cc)) {
          const k = nr * this.cols + nc;
          if (!this.revealed[k] && !this.mines[k]) {
            this.revealed[k] = 1;
            this.flagged[k] = 0;
            q.push([nr, nc]);
          }
        }
      }
    }
    return { result: 'safe', newly: this.safeRevealed - before };
  }

  toggleFlag(r, c) {
    const i = r * this.cols + c;
    if (this.over || !this.inBounds(r, c) || this.revealed[i]) return false;
    this.moves += 1;
    this.flagged[i] = this.flagged[i] ? 0 : 1;
    return true;
  }

  /** True if any neighbour of cell i is revealed (i is on the revealed frontier). */
  onFrontier(i) {
    const r = Math.floor(i / this.cols), c = i % this.cols;
    for (const [rr, cc] of this.neighbors(r, c)) if (this.revealed[rr * this.cols + cc]) return true;
    return false;
  }

  /** Move the mine at cell `from` to cell `to` and recompute every number. */
  moveMine(from, to) {
    if (!this.mines[from] || this.mines[to]) throw new Error('moveMine: no mine at source or mine at target');
    this.mines[from] = 0;
    this.mines[to] = 1;
    this._computeNumbers();
  }

  /** [r, c] of every cell that is neither revealed nor flagged. */
  hiddenCells() {
    const out = [];
    for (let r = 0; r < this.rows; r++) for (let c = 0; c < this.cols; c++) {
      const i = r * this.cols + c;
      if (!this.revealed[i] && !this.flagged[i]) out.push([r, c]);
    }
    return out;
  }
}
